File: application/helpers/svnupdate.php
<?php
/**
 * Este script permite ejecutar comandos de svn para desplegar un proyecto a
 * partir de un repositorio de subversion. Posteriormente adapta los ficheros config.php
 * y .htaccess para reconocer el subdirectio donde se ejecuta dentro del servidor web
 * y adaptar los permisos de la carpeta logs y templates_c para que puedan ser escritas
 * por el servidor web.
 * El funcionamiento de este script depende fuertemente del entorno de ejecucion y debera
 * ser adaptado a cada servidor.
 * 
 * svnupdate es un programa en c que debera tener los permisos adecuados (setuid) 
 * para ejecutar las acciones que no puede el usuario del servidor apache
 * ejemplos:
 * svn update
 * http://jair.lab.fi.uva.es/~marmari/svnupdate.php?p=<otramediapass>
 * checkout trunk 
 * http://jair.lab.fi.uva.es/~marmari/svnupdate.php?p=<otramediapass>&checkout=1
 * checkout specific branch
 * http://jair.lab.fi.uva.es/~marmari/svnupdate.php?p=<otramediapass>&checkout=1&branch=branches/codeIgniter
 */
echo "comienzo de la ejecucion<br />\n";
$usuarioJair = 'marmari';
$usuarioAssembla = 'mmarinero';
//es posible poner aqui la primera parte de la clave, el resto se debera pasar con el parametro p en la url.
$mediaPass = 'TlIunw0wbf';
$repositorio = 'https://subversion.assembla.com/svn/seteprosci/';
if (isset($_GET['branch'])) $repositorio .= $_GET['branch'];
else $repositorio .= 'trunk';

$configFile = 'application/config/config.php';
$htaccess = '.htaccess';
if (isset($_GET['checkout']) && $_GET['checkout']){
    exec('./svnupdate checkout /home/grini/'.$usuarioJair.'/.subversion '
            .$usuarioAssembla.' '.$mediaPass.escapeshellarg($_GET['p']).' '.$repositorio.' 2>&1',$svnOutput, $svnReturn);
} else {
    exec('./svnupdate update /home/grini/'.$usuarioJair.'/.subversion '
            .$usuarioAssembla.' '.$mediaPass.escapeshellarg($_GET['p']).' 2>&1',$svnOutput, $svnReturn);
}
if ($svnReturn) echo "El comando svn fallo, la salida se adjunta al final de la pagina<br />\n";
else echo "comando svn ejecutado correctamente<br />\n";

//Parece haber algun problema con file_get_contents aunque de momento funciona
//$fh = fopen($configFile, 'r');
//$CFstring = fread($fh, 32768);
//fclose($fh);
$CFstring = file_get_contents($configFile);
$replacedCFstring = str_replace("\$config['base_url']	= '';","\$config['base_url']	= '/~".$usuarioJair."/';",$CFstring);
$fp = popen('./svnupdate writefile config', 'w');
//pasamos por stdin el fichero a c
if (!fwrite($fp, $replacedCFstring)) echo 'no se pudo escribir '.$configFile."<br /> \n";
if(pclose($fp)) echo 'algo ha ido mal escribiendo el fichero '.$configFile." con svnupdate<br /> \n";
else echo 'Se ha modificado en '.$configFile." \$config['base_url'] = ''; por \$config['base_url'] = '/~".$usuarioJair."/';<br /> \n";

$HTstring = file_get_contents($htaccess);
$replacedHTstring = str_replace("RewriteBase /\n","RewriteBase /~".$usuarioJair."/\n",$HTstring);
$fp = popen('./svnupdate writefile htaccess', 'w');
if (!fwrite($fp, $replacedHTstring)) echo 'no se pudo escribir '.$htaccess."<br /> \n";
if(pclose($fp)) echo 'algo ha ido mal escribiendo el fichero '.$htaccess." con svnupdate<br /> \n";
else echo 'Se ha modificado en '.$configFile." RewriteBase / por RewriteBase /~".$usuarioJair."/<br /> \n";

foreach (array('logs', 'templates_c') as $file) {
    echo exec('./svnupdate chmod '.$file.' 2>&1',$chmodOutput, $chmodReturn);
    if ($chmodReturn) echo join("<br />\n",$chmodOutput)."<br />\n";
    else echo "$file chmoded correctamente <br />\n";
}
echo "<br />\nsalida svn:<br />\n&nbsp;&nbsp;&nbsp;&nbsp";
echo join("<br />\n&nbsp;&nbsp;&nbsp;&nbsp;",$svnOutput)."<br />\n";
echo "<br />\nfinal de ejecucion<br />\n";

File: application/helpers/svnupdate.c
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
int main(int argc, char **argv) {
    //update svn
    if (argc == 5  && 
        strcmp(argv[1], "update") == 0 && 
        strlen(argv[2]) < 100 && 
        strlen(argv[3]) < 100 && 
        strlen(argv[4]) < 100) {
        remove("application/config/config.php");
	remove(".htaccess");
        execl("/usr/bin/svn", "svn",
             "update", "--username",argv[3],
             "--password", argv[4],
             "--config-dir", argv[2],
             (const char *) NULL);
        fprintf(stderr, "Error al ejecutar svn update desde el ejecutable svnupdate\n");
        return(EXIT_FAILURE);
    }else if (argc == 6  &&
        strcmp(argv[1], "checkout") == 0 && 
        strlen(argv[2]) < 100 && 
        strlen(argv[3]) < 100 && 
        strlen(argv[4]) < 100 && 
        strlen(argv[5]) < 200) {
        int pid = fork();
        if (pid == 0) {
            execl("/usr/bin/find",".","-mindepth","1","-type","d","-exec","rm","-rf","{}","+",(const char *) NULL);
	} else {
            wait((int *) NULL);
            int pid2 = fork();
            if (pid2 == 0) {
                execl("/usr/bin/find",".","-type","f","!","-name","svnupdate*","-exec","rm","-f","{}","+",(const char *) NULL);
	    } else {
                wait((int *) NULL);
                execl("/usr/bin/svn", "svn", "checkout", argv[5], ".",
                     "--username",argv[3], "--password", argv[4], "--config-dir", argv[2], (const char *) NULL);
            }
        }
        fprintf(stderr, "Error al ejecutar svn checkout o rm desde el ejecutable svnupdate\n");
        return(EXIT_FAILURE);
    }else if (argc == 3  && strcmp(argv[1], "writefile") == 0) {
    //escribe fichero pasado por stdin si es uno de los posibles
 	char *filename;
        char buffer[32768];
        if (strcmp(argv[2], "config") == 0) {
            filename = "application/config/config.php";
	} else if (strcmp(argv[2], "htaccess") == 0) {
            filename = ".htaccess";
        } else {
            fprintf(stderr, "Fichero invalido\n");
            return(EXIT_FAILURE);
        }
        
        ssize_t nbytes;
        FILE *fp = fopen(filename, "w");

        if (fp == 0)
            error("No se pudo abrir el fichero\n");

        while ((nbytes = fread(buffer, sizeof(char), sizeof(buffer), stdin)) > 0)
            if (fwrite(buffer, sizeof(char), nbytes, fp) != nbytes)
                fprintf(stderr, "Error escribiendo fichero\n");
        fclose(fp);
    }else if (argc==3 && strcmp(argv[1],"chmod") == 0 && strlen(argv[2])<100) {
        char *filename;
        if (strcmp(argv[2], "logs") == 0) {
            filename = "application/logs";
	} else if (strcmp(argv[2], "templates_c") == 0) {
            filename = "application/views/templates_c";
        } else {
            fprintf(stderr, "Fichero invalido\n");
            return(EXIT_FAILURE);
        }
        execl("/bin/chmod","chmod","770",filename,(const char *) NULL);
        fprintf(stderr, "Error al ejecutar chmod\n");
        return(EXIT_FAILURE);
    } else {
        fprintf(stderr, "Parametros invalidos\n");
        return(EXIT_FAILURE);
    }
}
